documentation/modules/ROOT/pages/06-quarkus-reactive.adoc

```diff
@@ -0,0 +1,253 @@
+
+    @RestClient
+    BeerGateway beerGateway;
+
+    public Flowable<Beer> beers() {
+        Flowable<List<Beer>> beerStream = Flowable.generate(() -> 1, (page, emitter) -> { // <2>
+
+            beerGateway.getBeers(page).thenAcceptAsync(beers -> {
+                
+                    if (beers.isEmpty()) {
+                        emitter.onComplete(); // <3>
+                    } else {
+                        emitter.onNext(beers); // <4>
+                    }
+                
+                }, executor);
+
+                return page + 1; // <5>
+            });
+
+            return beerStream.flatMap(Flowable::fromIterable); // <6>
+    }
+
+}
+----
+copyToClipboard::qext-quarkusrpb-beer-service[]
+
+<1> Creates a single thread executor for getting beers by page
+<2> Gets first batch of beers
+<3> Sends an event that the process is completed
+<4> Emits each beer
+<5> Gets beers from next page
+<6> Transforms the list of beers event into event of beers
+
+So far nothing really new, but notice that now instead of reading all the beers, put them in memory and then filter out by any parameter, you are just emitting them individually per page.
+The caller is responsible to manipulate this information as it starts receiving the information and it is not necessary to store all beers in memory but just the ones that matches any condition.
+
+Notice that until a caller subscribes to a stream (`Flowable` in this case), the logic is not executed.
+
+[#quarkusrpb-beer-resource]
+==== Create Beer Resource
+
+The last step is to manipulate the beer _stream_ created in the previous section to filter out by _abv_ and return the result to the caller.
+
+Add new Java file in `$PROJECT_HOME/src/main/java/com/example` called `BeerResource` with the following contents:
+
+[#qext-quarkusrpb-beer-resource]
+[source,java,subs="+macros,+attributes"]
+----
+package com.example;
+
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.reactivestreams.Publisher;
+
+@Path("/beer")
+@Consumes(MediaType.APPLICATION_JSON)
+@Produces(MediaType.APPLICATION_JSON)
+public class BeerResource {
+
+    @Inject
+    BeerService beerService; // <1>
+
+    @GET
+    public Publisher<Beer> hello() {
+        return beerService.beers() // <2>
+                .filter(b -> b.getAbv() > 7.0); // <3>
+                // <4>
+    }
+}
+----
+copyToClipboard::qext-quarkusrpb-beer-resource[]
+
+<1> Injects `BeerService`
+<2> Gets the beer stream
+<3> For each beer emitted checks that _abv_ is greater than 7
+<4> Automatically subscribes to the stream and get back the result as `List<Beer>`
+
+The big change here is that now instead of having to deal with all beers in memory and then apply some operations, you are just loading into memory a _page_ of beers, then filter them and finally return them to the caller.
+
```

Comment:
Also, it will only call the beer service if the consumer request it (consume it). If the client abort before completion, you are going to save IO.